xapian-core/weight/tfidfweight.cc

```diff
@@ -63,16 +59,21 @@ TfIdfWeight::TfIdfWeight(const std::string &normals, double slope, double delta)
 	!strchr("ntpfsP", normalizations[1]) ||
 	!strchr("nP", normalizations[2]))
 	throw Xapian::InvalidArgumentError("Normalization string is invalid");
+    if (param_slope <= 0)
+	throw Xapian::InvalidArgumentError("Parameter slope is invalid.");
+    if (param_delta <= 0)
+	throw Xapian::InvalidArgumentError("Parameter delta is invalid.");
     if (normalizations[1] != 'n') {
 	need_stat(TERMFREQ);
 	need_stat(COLLECTION_SIZE);
     }
     need_stat(WDF);
     need_stat(WDF_MAX);
-    if (param_slope != 0 || param_delta != 0 || normalizations[2] == 'P') {
+    if ((param_slope > 0 || param_delta > 0) && normalizations[2] == 'P') {
```

Comment:
Now you're rejecting values `<= 0` in the constructor, `param_slope > 0` and `param_delta > 0` are now what's called "invariants" of this class.  If you want to check invariants, generally you use assertions, e.g.:

```
AssertRel(param_slope, >, 0);
AssertRel(param_delta, >, 0);
```

I'm not sure I'd bother here as you rejected the invalid values just above in the same function.  But more generally (e.g. when you're testing an invariant in different methods to where it was established) it can be useful as a defence against someone changing the code later and violating the invariant.

So you only need the final check here now.